logging.md

```diff
@@ -0,0 +1,109 @@
+
+### `human` output
+
+For `human` output, we expect a message like the following:
+```
+$ ./run-my-codemod.sh (...) /my/project
+[starting] 
+codemodder: java/1.2
+```
+
+Notice there is very little context for the log message. The log message doesn't show the timestamp, origin, thread, or any other contextual metadata. We believe CLI users would prefer a very simple, undecorated set of messages.
+
+### `json` output
+
+For `json` output, we expect a message like the following:
+
+```
+$ ./run-my-codemod.sh (...) /my/project
+{ "level" : "INFO", "message" : "[starting]" }
+{ "level" : "INFO", "message" : "codemodder: java/1.2" }
+```
+
+The level should be one of `DEBUG`, `INFO`, or `ERROR`. In general:
+ - `ERROR` messages should contain unexpected errors
+ - `INFO` should contain information about changes
+ - `DEBUG` should contain diagnostic or forensic information
+
+## Verbosity
+
+The `--verbose` flag tells the codemodder framework to emit all log messages, regardless of their level. This flag has nothing to do with the output format, and strictly controls the volume of messages emitted.
+
+## Output streams
+
+Generally, logs are expected to go to `stdout`. When the user specificies that the output format should be `human`, or the default case where no output format is specified, it's preferred that error messages be emitted to `stderr`.
+
+## Message Design
+
+We take inspiration from YAML as a format for designing log messages. Outputting in a YAML-like structure gives us some helpful strategies for how to visualize familiar structures, like lists of files, section breaks, and displaying simple key-value pairs.
+
+## Messages
+
+### Without `--verbose`
+
+The following example output shows all the required messages when running without `--verbose`. 
+
+```
+$ ./run-codemodder-codemod.sh --verbose -o /tmp/my_project.codetf /tmp/my_project
+codemodder: javascript/1.2
+
+[scanning]
+codemod: pixee:javascript/limit-readline
+changed:
+  - /tmp/my_project/file_d:57,59
+  - /tmp/my_project/file_e:194
+```
+
+### With `--verbose`
+
+The following example output shows all the required messages when running with `--verbose`. Frameworks can emit more messages, but the following messages serve as a reasonable first milestone or baseline for good-enough logging to support users and downstream tooling. Most of these messages feel self-explanatory, but the specification can go into more detail over time as actual elicits a clearer vision of what's important.
+
+```
+$ ./run-codemodder-codemod.sh --verbose -o /tmp/my_project.codetf /tmp/my_project
+
+[startup] 
+codemodder: javascript/1.2
+
+[setup]
+running: pixee:javascript/limit-readline, pixee:javascript/secure-random
+including paths: **/*.js
+excluding paths: **/tests/**,**/test.js
+
+[scanning]
+codemod: pixee:javascript/limit-readline
+scanned: 100 files...
+scanned: 200 files...
+scanned: 300 files...
+failed:
```

Comment:
I know this is just for the sake of example and it makes sense in that context, but in "real life" I would also expect to emit some error messages here related to the failures. I'm not sure how to call that out in the document though. I do think we need some language somewhere that says that codemodders are allowed to emit additional messages at various levels where appropriate (i.e. the spec doesn't strictly constrain the number/type of messages that can possibly be emitted).

Reply:
Doesn't this bit communicate this?

> Frameworks can emit more messages, but the following messages serve as a reasonable first milestone or baseline for good-enough logging to support users and downstream tooling.

Reply:
Yeah you're right, my bad.